[AArch64][SVE2] Generate XAR

UsmanNadeem opened the pull request:
Bitwise exclusive OR and rotate right by immediate

Select xar (x, y, imm) for the following pattern:
    or (shl (xor x, y), nBits-imm), (shr (xor x, y), imm)

This is essentially:
    rotr (xor(x, y), imm)

UsmanNadeem committed:
[AArch64][SVE2] Generate XAR

Bitwise exclusive OR and rotate right by immediate

Added a new ISD node for XAR and lower the following rotate pattern
to XAR for appropriate types:
    rotr (xor(x, y), imm) -> xar1 (x, y, imm)

Change-Id: If1f649b1bf5365b575dc9fa3e6618e97dc19a066

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -454,6 +454,15 @@ def AArch64eor3 : PatFrags<(ops node:$op1, node:$op2, node:$op3),
+  SDTCisVec<0>, SDTCisVec<1>, SDTCisVec<2>, SDTCisInt<3>,
+  SDTCisSameAs<0,1>, SDTCisSameAs<1,2>]>;
+def AArch64xar_node : SDNode<"AArch64ISD::XAR_I",  SDT_AArch64xar_Imm>;
+def AArch64xar : PatFrags<(ops node:$op1, node:$op2, node:$op3),
+                           [(int_aarch64_sve_xar node:$op1, node:$op2, node:$op3),
+                            (AArch64xar_node node:$op1, node:$op2, node:$op3)]>;
```
Since you're creating `XAR_I` to match the exact behaviour of the instruction we'd then typically extend `performIntrinsicCombine` to lower `int_aarch64_sve_xar` to `XAR_I` so that future combines catch the most cases and then you'll not need a PatFrags.

davemgreen commented:
It is better to avoid new DAG nodes if we can, as they can have some downsides and if we can match from the existing nodes that is generally beneficial. Can this be performed like the Neon instruction, using a method similar to trySelectXAR?

UsmanNadeem committed:
Revert "[AArch64][SVE2] Generate XAR"

This reverts commit b3e26b39271841a343e1f78b02ca17f69a3eb8ab.

UsmanNadeem committed:
[AArch64][SVE2] Generate XAR

Bitwise exclusive OR and rotate right by immediate

Select xar (x, y, imm) for the following pattern:
    or (shl (xor x, y), nBits-imm), (shr (xor x, y), imm)
This is essentially:
    rotr (xor(x, y), imm)

Change-Id: I55eac358745085e4f37c159ad3008113ac80a78b

UsmanNadeem committed:
fixup! [AArch64][SVE2] Generate XAR

UsmanNadeem commented:
Updated the patch to use code similar to how it is done in NEON.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -4275,6 +4275,64 @@ bool AArch64DAGToDAGISel::trySelectXAR(SDNode *N) {
+    uint64_t ShlAmt = LConst->getConstantOperandVal(0);
+    uint64_t ShrAmt = RConst->getConstantOperandVal(0);
+
+    if (ShlAmt + ShrAmt != VT.getScalarSizeInBits())
```
Is it worth having a negative test where R+L != Bitwidth?

UsmanNadeem replied:
Added!

davemgreen reviewed (commented):
Other than the comments below this LGTM, thanks for moving it.

UsmanNadeem committed:
Use isConstantSplatVector() and add couple more tests

Change-Id: I5852bd8ced920d9d3a4ab4b407ad309134f0f96d

UsmanNadeem commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -4275,6 +4275,64 @@ bool AArch64DAGToDAGISel::trySelectXAR(SDNode *N) {
+
+    SDValue LConst = N0.getOperand(2);
+    SDValue RConst = N1.getOperand(2);
+    if (RConst.getOpcode() != ISD::SPLAT_VECTOR ||
```
Done!

davemgreen reviewed (approved):
Thanks. LGTM
